src/main/java/com/es/phoneshop/service/cart/CustomCartService.java

```diff
@@ -0,0 +1,71 @@
+package com.es.phoneshop.service.cart;
+
+import com.es.phoneshop.exception.OutOfStockException;
+import com.es.phoneshop.model.attributesHolder.AttributesHolder;
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.model.cart.CartItem;
+import com.es.phoneshop.service.product.CustomProductService;
+import com.es.phoneshop.service.product.ProductService;
+
+import java.util.Optional;
+
+public class CustomCartService implements CartService {
+
+    private static final String CART_ATTRIBUTE_NAME = String.format("%s.cart", CustomCartService.class.getName());
+    private static volatile CustomCartService instance;
+    private ProductService productService;
+
+    private CustomCartService() {
+        productService = CustomProductService.getInstance();
+    }
+
+    public static CustomCartService getInstance() {
+        if (instance == null) {
+            synchronized (CustomCartService.class) {
+                if (instance == null) {
+                    instance = new CustomCartService();
+                }
+            }
+        }
+        return instance;
+    }
+
+    @Override
+    public Cart getCart(AttributesHolder attributesHolder) {
+        Cart cart = (Cart) attributesHolder.getAttribute(CART_ATTRIBUTE_NAME);
+        if (cart == null) {
+            cart = new Cart();
+            attributesHolder.setAttribute(CART_ATTRIBUTE_NAME, cart);
+        }
+        return cart;
+    }
+
+    @Override
+    public void addItem(Cart cart, Long productId, int quantity) throws OutOfStockException {
+        synchronized (cart) {
```

Comment:
maybe sync on session could be a small improvement? Usually CartModel is one of the most changeable instances in scope of e-commerce, so locking on this object may be less efficient